### ISSUE TITLE
[$1000]  Clicking the statement link in the Mac Desktop app opens the link in web instead of opening in the app sidebar directly

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#33977]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.22-0
**Reproducible in staging?:** needs reproduction
**Reproducible in production?:** needs reproduction
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @puneetlath 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1704219520043159

## Action Performed:
1. Click on the statement link from concierge chat in desktop app

## Expected Result:
Should open in the side bar

## Actual Result:
Opens in the web

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![image (7)](https://github.com/Expensify/App/assets/38435837/96e7ab77-fb11-49b8-a54b-00332f3cda59)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e81b7c4cb7cbfa0e</li>
        <li>Upwork Job ID: 1744433753657077760</li>
        <li>Last Price Increase: 2024-01-31</li>
<li>Automatic offers: </li>
<ul>
<li>akinwale | Reviewer | 28099226</li>
</ul></li>
    </ul>
</details>

**Comment by BriniM** [comment]
Contributor details
Your Expensify account email: maherbrini00@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/maherbrini

**Comment by bfitzexpensify** [comment]
Awaiting proposals.

**Comment by BriniM** [comment]
Is there a way to reproduce creating that Monthly statement, without having to enter any personal information, say as by having a test account or similar? Also I am fairly new to Expensify so a step-by-step guide / video recording on how to reproduce this would be nice.

**Comment by bfitzexpensify** [comment]
Hey @BriniM, does https://expensify.slack.com/archives/C01GTK53T8Q/p1701213684698809 answer your question?

**Comment by bfitzexpensify** [comment]
I'm going to be mostly ooo until Jan 29, assigning a second BZ team member to keep an eye on this.

- awaiting proposals

**Comment by Piotrfj** [comment]
Hello, Im Piotr from Callstack and would like to help with this issue

**Comment by MitchExpensify** [comment]
Sounds good @Piotrfj !

**Comment by MitchExpensify** [comment]
What's the latest on progress here @Piotrfj ?

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by MitchExpensify** [comment]
Friendly bump @Piotrfj

**Comment by Piotrfj** [comment]
Update: The issue is replicable and the source seems to be in link conversion

**Comment by MitchExpensify** [comment]
@bfitzexpensify Not much movement on this one while you were out. Worth doubling do you think?

**Comment by bfitzexpensify** [comment]
>Not much movement on this one while you were out. Worth doubling do you think?

Yep, I reckon - updated to $1000

**Comment by BriniM** [comment]
Wanted to reproduce the monthly statement but couldn't find a way to do this for a fresh expensify account with dummy data, would love to investigate this otherwise.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Clicking on a chat statement link on MacOS: Desktop / Native platforms opens the link in web instead of opening the RHN with the statement as it happens on Web / mWeb.

### What is the root cause of that problem?

The `RenderCommentHTML` component is responsible for rendering comments that have links in them, and in our case when there's a message with text + statement link (Concierge) or just a statement link, this is what the component renders:
- `<a href="https://staging.new.expensify.com/statements/202312" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/statements/202312</a>`

Because on MacOS: Desktop / Native platforms the we don't have logic to handle this type of external links with `target="_blank" rel="noreferrer noopener"` -> the app opens them via the browser.

### What changes do you think we should make in order to solve the problem?

**TLDR:** Use native / desktop specific component like `TextLink` which will handle the statement link internally.

A solution for this would require the following:
- create a library for both .desktop.ts and .native.ts, where index.ts would be noop
- use regex to check if the rendered html contains `statements/YYYYMM` format, like this: `/statements\/\d{6}/g`
- if any matches, next use regex to extract the href from the html, like this: `html.match(/href="([^"]*)/)?.[1]`
- finally, return `TextLink` component wrapping the statement link, making sure to also render any other text that might be part of the comment, if any

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Clicking on a chat statement link on MacOS: Desktop / Native platforms opens the link in web instead of opening the RHN with the statement as it happens on Web / mWeb.

### What is the root cause of that problem?

The `RenderCommentHTML` component is responsible for rendering comments that have links in them, and in our case when there's a message with text + statement link (Concierge) or just a statement link, this is what the component renders:
- `<a href="https://staging.new.expensify.com/statements/202312" target="_blank" rel="noreferrer noopener">https://staging.new.expensify.com/statements/202312</a>`

Because on MacOS: Desktop / Native platforms the we don't have logic to handle this type of external links with `target="_blank" rel="noreferrer noopener"` -> the app opens them via the browser.

### What changes do you think we should make in order to solve the problem?

**TLDR:** Use native / desktop specific component like `TextLink` which will handle the statement link internally.

A solution for this would require the following:
- create a library for both .desktop.ts and .native.ts, where index.ts would be noop
- use regex to check if the rendered html contains `statements/YYYYMM` format, like this: `/statements\/\d{6}/g`
- if any matches, next use regex to extract the href from the html, like this: `html.match(/href="([^"]*)/)?.[1]`
- finally, return `TextLink` component that onPress calls:
`Navigation.navigate(ROUTES.WALLET_STATEMENT_WITH_DATE.replace(':yearMonth', statementYearMonth))`
wrapping the statement link, making sure to also render any other text that might be part of the comment, if any

**Comment by ikevin127** [comment]
I think the problem comes from BE here and I came to this conclusion while debugging (more context above in my outdated proposal) - how so ? I noticed that there's no issues with any platform opening the RHN of the statement as long as the statement link has the following form:
- **DEV**: `https://dev.new.expensify.com:PORT/statements/202312` - the PORT match matters when testing locally
- **STG**: `https://staging.new.expensify.com/statements/202312`
- **PROD**: `https://new.expensify.com/statements/202312`

The only case when the app opens the link externally via the browser is when the statement's link domain env doesn't match the app's env which that's trying to open said link.

This is where I assume the BE sent the wrong env domain for the statement link via Concierge and when this issue was found, if it was found on staging, the link most likely had the prod env domain which the staging app doesn't recognize therefore it opens a new web tab regardless of the platform (web, desktop, natives).

**Note**: It doesn't have anything to do with the statement link being corrupt or anything like that - it's just the domain env mismatch with the app that's trying to open the link.

We can wait for confirmation from @m-natarajan as to how their Concierge statement link actually looks in terms of domain env and where the open attempt was performed.

For anyone who wants to test or confirm this, here's how you can generate a statement bypassing the wallet / bank flow:
- create a button somewhere in the UI and add the following function onClick / onPress:
`User.generateStatementPDF(yearMonth)` - yearMonth: string - The statement year and month as one string, i.e. 202110
- open Network tab to make sure it gets created when pressing the button (call succeeds)
- once you created a statement, this can be accessed via the links demonstrated above for each env
- you can just paste the link in a chat and click on it - notice that if the link matches your app's env / port (for dev) the statement RHN will open instead of a new browser tab and vice versa

**Comment by Piotrfj** [comment]
@MitchExpensify @m-natarajan Can you please provide a way to get the exact same message in Concierge chat?

**Comment by bfitzexpensify** [comment]
@m-natarajan can you clarify this? 

>We can wait for confirmation from @m-natarajan as to how their Concierge statement link actually looks in terms of domain env and where the open attempt was performed.

**Comment by bfitzexpensify** [comment]
Bump on this [question](https://github.com/Expensify/App/issues/33977#issuecomment-1928939072) @m-natarajan

**Comment by bfitzexpensify** [comment]
@puneetlath mind taking a look at [this comment](https://github.com/Expensify/App/issues/33977#issuecomment-1920321692) given you were the one who originally noticed this problem?

**Comment by ikevin127** [comment]
I think the statement links that Concierge sends always have prod env, no matter on what env the user is seeing said link.

Similar to the fact that when we're attempting login from staging / dev we always get the prod domain magic link (https://new.expensify.com/...). More context about this here https://github.com/Expensify/App/issues/34177#issuecomment-1927678535.

Technically we could have a FE based fix which would always make sure that we convert all expensify domain links to the current env of the user, but I think this should be handled by BE in all cases.

**Comment by bfitzexpensify** [comment]
Thanks @ikevin127 

cc @thienlnam from the work in https://github.com/Expensify/Expensify/issues/213310 - Jack, mind taking a look at https://github.com/Expensify/App/issues/33977#issuecomment-1920321692 and https://github.com/Expensify/App/issues/33977#issuecomment-1936025590 and confirming you think this should be a BE fix?

**Comment by bfitzexpensify** [comment]
Mind confirming this @puneetlath? 

>I'm guessing Puneet might have opened this from the staging desktop environment and then clicked on the link which opened prod in web?

If that's what happened, then we can close this out.

**Comment by puneetlath** [comment]
Ah yeah, that's probably what it was. I'll try it again when the next statement comes. Thanks y'all!
